Add compressed-display-math and include-characters

slotThe opened the pull request:
As promised, this adds two new variables, `math-delimiters-compressed-display-math` and `math-delimiters-include-characters`.  Very briefly, the former is for toggling to a more liberal math display mode (with regards to newlines) and the latter slurps and barfs the given characters when going into and out of math display mode, respectively.

The commits should hopefully be self-explanatory, so I will let them do most of the talking.

#### Commit Summary

- Escape `\[` in documentation

  As per the elisp manual [1], sequences of the form `\\[blah]` interpret
  `blah` as a command and are substituted either by key sequences, or by
  `M-x blah`, should there be no valid sequence.  Fix this by inserting a
  forced quote.  This also fixes a small typo.

  [1]: https://ftp.gnu.org/old-gnu/Manuals/elisp-manual-21-2.8/html_chapter/elisp_24.html#SEC376

- Add math-delimiters-include-characters

  Adds a new variable, `math-delimiters-include-characters`.  When
  converting from inline to display math, this specifies a list of
  characters to include in the display math environment.  When converting
  the other way around, they are excluded again.

- Add math-delimiters-compressed-display-math

  Adds a new variable, `math-delimiters-compressed-display-math`, to
  toggle surrounding the display math delimiters with newlines, resulting
  in a more liberal representation.

  The old (and now default) behaviour was going from

      text-before \(1 + 1\) text-after

  to

      text-before \[1 + 1\] text-after.

  Setting this value to `nil` switches between the above inline math
  version and

      text-before
      \[
        1 + 1
      \]
      text-after.

- Update documentation: compressed-display-math and include-characters

EDIT: One more thing I noticed while trying this out: the modes in the `with-eval-after-load` statements are unquoted 

``` emacs-lisp
(with-eval-after-load org
  (define-key org-mode-map "$" #'math-delimiters-insert))
```

but I don't think this work (at least it doesn't for me).  We probably need to either quote this or write it as a string

slotThe committed:
Escape \[ in documentation

As per the elisp manual [1], sequences of the form "\\[blah]" interpret
`blah` as a command and are substituted either by key sequences, or by
`M-x blah`, should there be no valid sequence.  Fix this by inserting a
forced quote.  This also fixes a small typo.

[1]: https://ftp.gnu.org/old-gnu/Manuals/elisp-manual-21-2.8/html_chapter/elisp_24.html#SEC376

slotThe committed:
Add math-delimiters-include-characters

Adds a new variable, `math-delimiters-include-characters`.  When
converting from inline to display math, this specifies a list of
characters to include in the display math environment.  When converting
the other way around, they are excluded again.

slotThe committed:
Add math-delimiters-compressed-display-math

Adds a new variable, `math-delimiters-compressed-display-math`, to
toggle surrounding the display math delimiters with newlines, resulting
in a more liberal representation.

The old (and now default) behaviour was going from

    text-before \(1 + 1\) text-after

to

    text-before \[1 + 1\] text-after.

Setting this value to `nil` switches between the above inline math
version and

    text-before
    \[
      1 + 1
    \]
    text-after.

slotThe committed:
Update documentation: compressed-display-math and include-characters

slotThe commented on math-delimiters.el:
```diff
@@ -85,12 +100,88 @@
+`math-delimiters-include-characters'.  Thus, include these
+characters in display math should they come right after an inline
+math statement, and exclude them in the other direction."
+  (cl-flet ((on-char-p (pos)
+                       (cl-loop for punct in math-delimiters-include-characters
+                                do (when (= punct pos)
+                                     (cl-return t)))))
```
This is a big awkward; what I really want is to write this in one line with a function of the form `any: ∀a. Hom(a, Bool) × [a] → Bool` (to write it mathematically because we're talking TeX here :)) but I could not find anything like that outside of `dash.el`

oantolin replied:
Well, here I'd simply use `(memq pos math-delimiters-include-characters)`, but there are functions like `any` in Emacs: `seq-find` and `cl-find-if`. (I don't think dash actually has much that isn't built-in, my impression is that people like it because they prefer the names it uses, not because Emacs doesn't have that functionality.)

oantolin replied:
I changed this to use `cl-find` and made `math-delimiters-include-characters` a string instead of a list, since it feels easier to type `",.;"` than `'(?, ?. ?;)`.

slotThe replied:
> Well, here I'd simply use (memq pos math-delimiters-include-characters), but there are functions like any in Emacs: seq-find and cl-find-if.

Oh that's really neat, thanks!  I suppose I should finally take a gander at the elisp manual after all

slotThe commented on math-delimiters.el:
```diff
@@ -85,12 +100,88 @@
+              (math-delimiters-compressed-display-math
+               (goto-char orig-pos)))))))
+
+(defun math-delimiters--toggle-line-breaks (open close)
+  (let* ((to-display (equal (car math-delimiters-display) open))
+         (toggle-closing `(if ,to-display
+                              (progn
+                                (TeX-newline)
+                                (search-backward ,close)
+                                (TeX-newline))
+                            (join-line -1)
+                            (join-line -1)
+                            (backward-char (1+ (length ,close)))))
+         (toggle-open `(if ,to-display
+                           (progn (TeX-newline)
+                                  (forward-char (length ,open))
+                                  (TeX-newline))
+                         (join-line)
+                         (join-line -1)
+                         ;; If necessary, like when the inline delimiters are
+                         ;; Dollars, fix `join-line's whitespace insertion.
+                         (when (equal (char-after) ?\s)
+                           (delete-char 1)))))
+    (eval toggle-closing)
+    (search-backward open)
+    (eval toggle-open))
+  (search-forward close))
```
This is a bit spaghetti, but there was no good way I could see to make this more generic and not differentiate betwen inline -> display and display -> inline here

oantolin replied:
We definitely don't need `eval` for this, I'll rewrite it a little differently.

oantolin replied:
Oh, we can't use `TeX-newline`, that's an AUCTeX function. I've been careful to have the package also useable with the built-in tex modes.

oantolin replied:
I fixed this part to use local functions instead of `eval` and to use the neutral `newline-and-indent`.

slotThe replied:
Oh I saw "TeX-newline is [...] defined in tex.el" and didn't investigate further because surely auctex wouldn't use a file called that when there is a built in tex mode; guess I was wrong :)  Thanks for fixing these issues

slotThe commented on math-delimiters.el:
```diff
@@ -129,14 +220,17 @@ delimeters."
                               (goto-char (+ end (- (length to-open)
-                                                   (length from-open))))))
+                                                   (length from-open)))))
+                            (unless math-delimiters-compressed-display-math
+                              (math-delimiters--toggle-line-breaks to-open to-close)))
```
When using the uncompressed display math option, we are first doing the compressed case anyways and then post-process things.  We could also differentiate a bit sooner, but this would require proper extra functions for the compressed and uncompressed version.  This has the advantage that `math-delimiters--slurp-or-barf-characters` is _almost_ compeltely unaware of this distinction and it results in (marginally) prettier code.

I don't think I notice any speed difference between these two options.  I have kept a (hopefully working) version of the other approach around; in case you strongly prefer that approach I can try to replace the current implementation with that

slotThe reviewed (commented):
All of this comes with the huge caveat that I don't really know what I'm doing with regards to elisp (and especially with regards to `cl-lib`), so proceed with caution :)

oantolin commented:
Thank you for doing this! I'll take a look at the code now, and then respond to the comments you wrote above, but right off the bat I'm wondering if we really need the compressed display math option? Maybe we could just keep the new behavior and get rid of the old compressed one? I'm not sure, maybe some people do want compressed display math. Are there some standard guideline for formatting LaTeX source? If adding newlines is the recommended approach, I'd feel justified in only offering the newline-y behavior.

oantolin commented:
There are a couple of problems left in the case of empty delimiters in uncompressed mode:

- if you toggle from inside the delimiter you don't get the newlines added
- if you toggle from outside the delimiters there is a failing search and the cursor gets felt in the wrong spot.

oantolin committed:
Define local functions instead of using eval

oantolin committed:
Make math-delimiters-include-characters a string instead

This is a more convenient way to specify a set of characters.

Also, change the default value from no characters to comma, period and
semicolon, which I think pretty much everyone will want.

oantolin committed:
Fix cl-flet indentation

oantolin committed:
Fix bug when toggling empty delimiters from outside

oantolin commented:
I fixed the bug when toggling empty delimiters from the outside. The only thing left that I can see is to handle adding the newlines in uncompressed mode when toggling empty delimiters.

oantolin committed:
Respect math-delimiters-compressed-display-math from middle

oantolin commented:
I think that's it! Ready to merge. Thanks a lot for doing this!

slotThe commented:
> right off the bat I'm wondering if we really need the compressed display math option? Maybe we could just keep the new behavior and get rid of the old compressed one? I'm not sure, maybe some people do want compressed display math. Are there some standard guideline for formatting LaTeX source? If adding newlines is the recommended approach, I'd feel justified in only offering the newline-y behavior.

I'm sure you know as well as I do how... creative people can get with formatting LaTeX files :)  I haven't stumbled upon an actual styles guide beyond just folk wisdom and Knuth himself seems to use the "compressed" style [quite a lot](https://www-cs-faculty.stanford.edu/~knuth/err1.textxt) so it was probably the right decision to leave it in.  At the end of the day, the introduced flexibility means that it's one less thing to be hung up on when working with that one coauthor who _really_ insists on their notation

oantolin commented:
> At the end of the day, the introduced flexibility means that it's one less thing to be hung up on when working with that one coauthor who _really_ insists on their notation

True!
